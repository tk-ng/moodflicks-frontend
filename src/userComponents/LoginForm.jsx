import { useFields } from "../hooks/useFields";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { useNavigate } from "react-router";
import { toast } from "react-toastify";

export const LoginForm = ({ login }) => {
	const navigate = useNavigate();

	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await login(formData);
			resetForm();
			navigate("/");
			toast.success(`Welcome back, ${formData.username}!`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="LoginForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Log In</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
