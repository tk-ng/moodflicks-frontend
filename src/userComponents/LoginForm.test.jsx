import { render, screen } from "@testing-library/react";
import { LoginForm } from "./LoginForm";
import { MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";

describe("Render the LoginForm component", function () {
	it("mounts without crashing", async function () {
		const { container, asFragment } = render(
			<MemoryRouter>
				<LoginForm />
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders the LoginForm properly", async function () {
		render(
			<MemoryRouter>
				<LoginForm />
			</MemoryRouter>
		);
		expect(screen.getByLabelText("Username")).toBeInTheDocument();
		expect(screen.getByLabelText("Password")).toBeInTheDocument();
		expect(
			screen.getByText("Submit", { selector: "button" })
		).toBeInTheDocument();
	});
});

describe("Submit the LoginForm", function () {
	it("submits the form data properly", async function () {
		const login = vi.fn();
		render(
			<MemoryRouter>
				<LoginForm login={login} />
			</MemoryRouter>
		);
		const user = userEvent.setup();
		await user.type(screen.getByLabelText("Username"), "testuser1");
		await user.type(screen.getByLabelText("Password"), "testpassword");
		await user.click(screen.getByText("Submit", { selector: "button" }));
		expect(login).toHaveBeenCalledWith({
			username: "testuser1",
			password: "testpassword",
		});
	});
});
