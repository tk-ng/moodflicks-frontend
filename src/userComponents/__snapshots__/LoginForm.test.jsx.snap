// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render the LoginForm component > mounts without crashing 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="row"
    >
      <div
        class="container col-sm-12 col-md-6 offset-md-3 col-lg-4 offset-lg-4"
      >
        <h3
          class="mb-3"
        >
          Log In
        </h3>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <form
              class=""
            >
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="username"
                >
                  Username
                </label>
                <input
                  class="form-control"
                  id="username"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="form-control"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
              </div>
              <div
                class="d-grid"
              >
                <button
                  class="btn btn-primary"
                >
                  Submit
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
